keps/sig-network/4974-deprecate-endpoints/README.md

```diff
@@ -0,0 +1,659 @@
+to modify them. (This might be problem with either "real" or "virtual"
+annotations.)
+
+#### Warnings in `kubectl get endpoints`
+
+We could potentially have `kubectl` print warnings (to stderr) when
+anyone gets or lists endpoints. There doesn't seem to be precedent for
+this, but it seems like it should work.
+
+#### Warnings via logs/metrics/Events
+
+We could potentially warn within the apiserver about clients Getting,
+Listing, or Watching Endpoints objects. (This would presumably be
+generic and based on the fact that the v1.Endpoints type is marked as
+deprecated. Maybe there is already something like this?)
+
+### `kubernetes.default` Endpoints
+
+The Endpoints for the `kubernetes.default` Service are created by the
+apiserver rather than by kube-controller-manager (and are thus
+independent of whether the Endpoints controller is running or not).
+
+We have a few options here:
+
+  1. Do nothing, continue to always generate both an Endpoints and
+     an EndpointSlice for `kubernetes.default` in all clusters.
+
+  2. As with #1, but add a deprecation warning annotation to the
+     Endpoints object.
+
+  3. Add a config option to kube-apiserver to optionally disable
+     `kubernetes.default` Endpoints generation (possibly with the
+     warning as in #2 in the case where it's not disabled).
+
+  4. Do #1, #2, or #3, but eventually (maybe a long time in the
+     future) unconditionally stop generating the Endpoints object.
+
+Whether #4 is allowed or not depends on whether we think there is an
+"API"-type guarantee that a `kubernetes.default` Endpoints object
+exists (or if the "API" is merely that the `kubernetes.default`
+_Service_ exists and "works", without reference to _how_ it works).
+
+```
+<<[UNRESOLVED kubernetes.default ]>>
+
+Pick an option.
+
+<<[/UNRESOLVED]>>
+```
+
+### Endpoints Cleanup
+
+We do not want to leave stale `Endpoints` objects around forever if
+the Endpoints controllers are not running. (This is both a waste of
+disk space and a potential source of confusion since the Endpoints
+objects would quickly become out-of-date and incorrect.)
+
+One possibility would be to just document that administrators should
+delete all existing Endpoints themselves if they are going to disable
+the controllers.
+
+Another possibility would be to have kube-controller-manager do this
+automatically. (Though the idea of having it automatically take action
+because a controller _isn't_ enabled seems slightly magical? But
+perhaps there is precedent somewhere else?)
+
+Another possibility would be to create an `endpoints-cleanup`
+controller, that could be enabled explicitly, and document that admins
+should (probably) enable that controller if they disable the others.
+
+Or perhaps the EndpointSlice controller could delete Endpoints objects
+that were more than 24 hours out of date with respect to their
+EndpointSlices?
+
+In all cases, we should probably not automatically delete Endpoints
```

Comment:
Do we track that?  I don't think we have an annotation that says so?  Maybe we should add that NOW so we can use it later?

Reply:
"Endpoints that were not originally created by the Endpoints controller" is not _exactly_ the right concept, I guess.

We should delete all Endpoints that the Endpoints controller itself would have updated/deleted under appropriate circumstances (ie, any Endpoints that corresponds to a Service with a selector). We should not delete Endpoints that the Endpoints controller would have ignored (ie, Endpoints that do not correspond to a Service or that correspond to a selectorless Service).